### ISSUE TITLE
XMLHttpRequest: "responseType" is not being honored when making non-mocked requests

### ISSUE THREAD

**Issue by avm99963** [mswjs/interceptors#309]
When running the following code in a browser:

``` js
import {XMLHttpRequestInterceptor} from '@mswjs/interceptors/lib/interceptors/XMLHttpRequest';

const URL = 'https://httpbin.org/get';

const interceptor = new XMLHttpRequestInterceptor();
interceptor.apply();

// Send a request expecting to receive an ArrayBuffer.
const xhr = new XMLHttpRequest();
xhr.responseType = 'arraybuffer';
xhr.open('GET', URL);

console.log('xhr.responseType (original):', xhr.responseType);

xhr.addEventListener('load', () => {
  console.log('xhr.responseType (new):', xhr.responseType);
  console.log('xhr.response:', xhr.response);
});

xhr.send();
```

The following output is shown in the JS console:

```
xhr.responseType (original): arraybuffer

xhr.responseType (new): 

xhr.response: {
  "args": {}, 
  "headers": {
    [censored]
  }, 
  "origin": "[censored]", 
  "url": "https://httpbin.org/get"
}
```

I'd expect `xhr.responseType` to continue being `arraybuffer` and `xhr.response` to be an ArrayBuffer.

PS: Thanks for developing this! It's awesome :) I'm trying to use it for [a Chrome extension](https://gerrit.avm99963.com/plugins/gitiles/infinitegforums/) which needs to modify some responses to improve how a website is shown.

**Comment by avm99963** [comment]
See my attempt to fix this here: https://github.com/mswjs/interceptors/compare/main...avm99963:interceptors:b/309

I added a test to `test/modules/XMLHttpRequest/intercept/XMLHttpRequest.test.ts`, but it fails with

> `Error: Uncaught [InvalidStateError: The object is in an invalid state.]`

due to the added line in `src/interceptors/XMLHttpRequest/XMLHttpRequestOverride.ts` (which I think would fix this).

**Comment by kettanaito** [comment]
Thanks for reporting this, @avm99963! 

It makes me wonder if this logic makes sense: 

https://github.com/mswjs/interceptors/blob/ddf0389c0a0f2f98350f3857d1e301f806c66587/src/interceptors/XMLHttpRequest/XMLHttpRequestOverride.ts#L419-L423

We're inferring the `responseType` from the original request onto the patched instance. I guess it should be the other way around unless I'm missing something. `responseType` is something set when defining the XMLHttpRequest, the actual server response has no control over it. 

I think that's the root cause. To put it briefly:

1. The logic I linked above must be removed;
2. Instead, we should make sure we infer all the properties that were configured on the patched instance onto the original request instance (e.g. `responseType`, `withCredentials`, etc);
3. Write automates tests to cover those inferred properties when performing a bypass request.

**Comment by kettanaito** [comment]
@avm99963, would you be interested in trying those suggestions of mine? I can land you a hand on code review.

**Comment by avm99963** [comment]
Hey @kettanaito, thanks for the quick reply! :)

By reading the [definition of `responseType`](https://xhr.spec.whatwg.org/#the-responsetype-attribute) in the XMLHttpRequest standard (and taking a look at each section that references `responseType`), I think what you say is true: it's never changed by the user agent, it only changes when the Javascript code sets it. So LGTM to remove those lines.

I'll definitely do what you outlined. I'll also check whether the user agent can change any of the other properties in case we have to implement something similar to the lines we're removing for other properties.

Thanks!

**Comment by avm99963** [comment]
Done, it seems like we already took into account the `timeout` property, so in the end I just added the logic for the `responesType` and `withCredentials` properties. I forced push the new commit to the branch I previously shared, and created a PR so we can start the code review there.

However, I'm not sure of how to approach the tests. What's a bypass request? Is it when the actual request is made instead of returning a mocked response?

**Comment by kettanaito** [comment]
I think the user agent cannot really change any request properties in XMLHttpRequest. The intention behind the current logic was wrong, I think it should've been the other way around: infer mocked request properties onto the original (bypassed) request. I also believe this is how it was previously. 

Thanks for working on this! I will help you with tests so we're sure this issue won't happen again.

**Comment by kettanaito** [comment]
## Released: v0.20.0 🎉

This has been released in v0.20.0!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.20.0)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.20.0)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
